packages/blocks-to-image/src/index.ts

```diff
@@ -0,0 +1,146 @@
+import { KnownBlock } from '@slack/types';
+import puppeteer, {
+  BoundingBox,
+  BoxModel,
+  Browser,
+  LaunchOptions,
+  Page,
+  ScreenshotOptions,
+} from 'puppeteer';
+import querystring from 'querystring';
+
+type Screenshot = string | Buffer;
+
+enum BlockSurfaceModeSelectors {
+  'message' = '.p-block_kit_builder_preview__message',
+  'modal' = '.p-block_kit_builder_preview_modal',
+  'appHome' = '.p-block_kit_builder_preview__app_home',
+}
+
+type BlockSurfaceModes = keyof typeof BlockSurfaceModeSelectors;
+
+const minSlackViewportWidth = 1000; // Magic number, width needed for builder to render
+
+async function setViewportFromClip(
+  page: Page,
+  width: number,
+  clip: BoundingBox,
+): Promise<void> {
+  await page.setViewport({
+    width: Math.ceil(width),
+    height: Math.ceil(clip.height + clip.x),
+    deviceScaleFactor: 1,
+  });
+}
+
+async function screenshotDOMElement(
+  page: Page,
+  selector: string,
+  options?: ScreenshotOptions,
+): Promise<Screenshot> {
+  const element = await page.waitForSelector(selector, { visible: true });
+  // Removing the possibility of `null`
+  const elementBox = (await element.boxModel()) as BoxModel;
+
+  const marginTopLeft = elementBox.margin[0];
+  const marginTopRight = elementBox.margin[1];
+  const marginBottomRight = elementBox.margin[3];
+
+  const screenshotOptions = {
+    clip: {
+      x: marginTopLeft.x,
+      y: marginTopLeft.y,
+      width: marginTopRight.x - marginTopLeft.x,
+      height: marginBottomRight.y - marginTopRight.y,
+    },
+  };
+
+  await setViewportFromClip(
+    page,
+    minSlackViewportWidth,
+    screenshotOptions.clip,
+  );
+
+  await page.waitForFunction(
+    `document.querySelector('${selector}') && document.querySelector('${selector}').clientHeight != 0`,
+  );
+
+  const screenshot = await page.screenshot({
+    ...options,
+    ...screenshotOptions,
```

Comment:
user couldn't override `clip` option given this arrangement, should we swap these?

Reply:
I did this purposefully because we're spending a lot of time calculating the clip to get _just_ the element. As we don't expose a way to get a different clip, it seemed weird to me to make it an option. CA: 1